src/fairseq2/datasets/huggingface.py

```diff
@@ -0,0 +1,191 @@
+        )
+
+
+def create_hf_reader(
+    dataset: Dataset,
+    gang: Gang,
+    converter: Callable[[Example], Seq2SeqBatch],
+    *,
+    batcher: Optional[Batcher] = None,
+    num_accumulate: int = 1,
+    num_prefetch: int = 1,
+    pad_value: Optional[int] = None,
+    **extra: Any,
+) -> DataPipelineReader[Seq2SeqBatch]:
+    """
+    Convert HF Dataset into a fairseq2 PipelineReader.
+    The HF-specific processing logic is assumed to be
+    done before the conversion.
+
+    **Batching**: The batching can be done externally or using
+    bucketing strategies in fairseq2. If both ``bucket_size``
+    and ``max_num_elements`` are None, then we assume the
+    ``dataset`` is already batched and do nothing. Otherwise:
+
+    - If ``bucket_size`` is not None, generate batches of fixed size
+    - If ``max_num_elements`` is not None, generate batches up to
+    `max_num_elements` size. In this case, ``max_seq_len`` must be
+    given.
+
+    In addition, the data
+    should be prepared to have a column that contains
+    sequence length for each example.
+
+    This function should not be used for big datatset.
+
+    :param gang:
+        The gang over which to shard the dataset.
+    :param batcher:
+        The batching strategy that is used to transform the pipeline
+        into iteration of batch of items.
+    :param num_accumulate:
+        The number of batches to accumulate in each iteration.
+        Typically used with gradient accumulation during
+        training.
+    :param num_prefetch:
+        The number of batches to prefetch in background.
+    :param extras:
+        The extra parameters specific to the dataset
+        implementation.
+    """
+
+    if not has_datasets:
+        raise ModuleNotFoundError(
+            "`datasets` is required but not found. Please install it with `pip install datasets`."
+        )  # fmt: skip
+
+    # Make sure the dataset is a proper arrow dataset
+    if not isinstance(dataset, Dataset):
+        # One common mistake is pass a DatasetDict (e.g. with all splits) as inputs
+        if isinstance(dataset, DatasetDict):
+            raise TypeError(
+                "create_reader() expects input of datasets.Dataset type, "
+                "get datasets.DatasetDict. Make sure you specify `split` "
+                "when loading the dataset"
+            )
+        raise TypeError(f"Expect datasets.Dataset type, get {type(dataset)}")
+
+    if batcher is None:
+        data = dataset.data.to_batches()  # list of RecordBatch
+    else:
+        data = dataset
+    builder = read_sequence(data)
+
+    # Shard.
+    builder.shard(gang.rank, gang.size, allow_uneven=True)
+
+    if batcher is None:
+        # Convert RecordBatch to python dictionary
+        builder = builder.map(lambda batch: batch.to_pydict())
+    else:
+        builder = batcher.batch(builder, **extra)
+
+        # collate to python dict
+        builder.map(Collater(pad_value=pad_value))
```

Comment:
Updated `builder.map(Collater())` to `builder.map(Collater(pad_value=pad_value))` because the collater will only stack sequences unless a pad_value is specified (default is None).